Missing instantiations for `TrilinosWrappers::SparseMatrix::vmult` when compiling a user project linked against deal.II compiled with clang

When compiling the deal.II library itself with clang and a user project with gcc I run into a missing instantiation error:
```
mold: error: undefined symbol:
std::enable_if<is_same_v<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::value_type, double>, void>::type
dealii::TrilinosWrappers::SparseMatrix::vmult<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> >(
    dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, 
    dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&) const
```

I can trigger a similar link error by compiling the user project with clang as well when switching to float:
```
mold: error: undefined symbol:
std::enable_if<is_same_v<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::value_type, double>, void>::type
dealii::TrilinosWrappers::SparseMatrix::vmult<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> >(
    dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&,
    dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&) const
```

The `vmult()` function is templated with `template <typename VectorType>`, so we must explicitly instantiate it. Judging from `source/lac/trilinos_sparse_matrix.cc` and `*.inst.in` it seems that we never explicitly instantiate the `vmult()` function?

Am I mistaken?

Hi all,

I encountered a similar issue while compiling ASPECT with deal.II, both using the **Clang** compiler. The error is:

```md
undefined reference to 'std::array<std::pair<double, dealii::Tensor<1, 3, double> >, std::integral_constant<int, 3>::value> dealii::eigenvectors<3, double>(dealii::SymmetricTensor<2, 3, double> const&, dealii::SymmetricTensorEigenvectorMethod)'
```

It seems that the template function `dealii::eigenvectors<3, double>` is not instantiated for the specific types being used. Based on this issue's discussion, I suspect an explicit instantiation might be missing, similar to the problem with `TrilinosWrappers::SparseMatrix::vmult`.

Would an explicit instantiation of `dealii::eigenvectors<3, double>` solve this?

Thanks for any advice!

--- a/include/deal.II/base/symmetric_tensor.h
+++ b/include/deal.II/base/symmetric_tensor.h
@@ -3025,8 +3025,7 @@ enum struct SymmetricTensorEigenvectorMethod
  * @relatesalso SymmetricTensor
  */
 template <int dim, typename Number>
-std::array<std::pair<Number, Tensor<1, dim, Number>>,
-           std::integral_constant<int, dim>::value>
+std::array<std::pair<Number, Tensor<1, dim, Number>>, dim>
 eigenvectors(const SymmetricTensor<2, dim, Number> &T,
              const SymmetricTensorEigenvectorMethod method =
                SymmetricTensorEigenvectorMethod::ql_implicit_shifts);

--- a/include/deal.II/base/symmetric_tensor.templates.h
+++ b/include/deal.II/base/symmetric_tensor.templates.h
@@ -848,8 +848,7 @@ namespace internal
 
 
 template <int dim, typename Number>
-std::array<std::pair<Number, Tensor<1, dim, Number>>,
-           std::integral_constant<int, dim>::value>
+std::array<std::pair<Number, Tensor<1, dim, Number>>, dim>
 eigenvectors(const SymmetricTensor<2, dim, Number> &T,
              const SymmetricTensorEigenvectorMethod method)
 {
@@ -1055,18 +1054,18 @@ eigenvalues(const SymmetricTensor<2, 3, adouble> & /*T*/);
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 1, adouble>>, 1>
-eigenvectors(const SymmetricTensor<2, 1, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/);
+eigenvectors<1, adouble>(const SymmetricTensor<2, 1, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/);
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 2, adouble>>, 2>
-eigenvectors(const SymmetricTensor<2, 2, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/);
+eigenvectors<2, adouble>(const SymmetricTensor<2, 2, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/);
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 3, adouble>>, 3>
-eigenvectors(const SymmetricTensor<2, 3, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/);
+eigenvectors<3, adouble>(const SymmetricTensor<2, 3, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/);
 #endif
 
 DEAL_II_NAMESPACE_CLOSE

--- a/source/base/symmetric_tensor.cc
+++ b/source/base/symmetric_tensor.cc
@@ -83,8 +83,8 @@ eigenvalues(const SymmetricTensor<2, 3, adouble> & /*T*/)
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 1, adouble>>, 1>
-eigenvectors(const SymmetricTensor<2, 1, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/)
+eigenvectors<1, adouble>(const SymmetricTensor<2, 1, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/)
 {
   AssertThrow(false, ExcADOLCAdvancedBranching());
   return std::array<std::pair<adouble, Tensor<1, 1, adouble>>, 1>();
@@ -94,8 +94,8 @@ eigenvectors(const SymmetricTensor<2, 1, adouble> & /*T*/,
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 2, adouble>>, 2>
-eigenvectors(const SymmetricTensor<2, 2, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/)
+eigenvectors<2, adouble>(const SymmetricTensor<2, 2, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/)
 {
   AssertThrow(false, ExcADOLCAdvancedBranching());
   return std::array<std::pair<adouble, Tensor<1, 2, adouble>>, 2>();
@@ -105,8 +105,8 @@ eigenvectors(const SymmetricTensor<2, 2, adouble> & /*T*/,
 
 template <>
 std::array<std::pair<adouble, Tensor<1, 3, adouble>>, 3>
-eigenvectors(const SymmetricTensor<2, 3, adouble> & /*T*/,
-             const SymmetricTensorEigenvectorMethod /*method*/)
+eigenvectors<3, adouble>(const SymmetricTensor<2, 3, adouble> & /*T*/,
+                         const SymmetricTensorEigenvectorMethod /*method*/)
 {
   AssertThrow(false, ExcADOLCAdvancedBranching());
   return std::array<std::pair<adouble, Tensor<1, 3, adouble>>, 3>();
@@ -124,16 +124,16 @@ template std::array<adouble, 3>
 eigenvalues(const SymmetricTensor<2, 3, adouble> &);
 
 template std::array<std::pair<adouble, Tensor<1, 1, adouble>>, 1>
-eigenvectors(const SymmetricTensor<2, 1, adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<1, adouble>(const SymmetricTensor<2, 1, adouble> &,
+                         const SymmetricTensorEigenvectorMethod);
 
 template std::array<std::pair<adouble, Tensor<1, 2, adouble>>, 2>
-eigenvectors(const SymmetricTensor<2, 2, adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<2, adouble>(const SymmetricTensor<2, 2, adouble> &,
+                         const SymmetricTensorEigenvectorMethod);
 
 template std::array<std::pair<adouble, Tensor<1, 3, adouble>>, 3>
-eigenvectors(const SymmetricTensor<2, 3, adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<3, adouble>(const SymmetricTensor<2, 3, adouble> &,
+                         const SymmetricTensorEigenvectorMethod);
 #  endif
 
 template std::array<adtl::adouble, 1>
@@ -146,16 +146,16 @@ template std::array<adtl::adouble, 3>
 eigenvalues(const SymmetricTensor<2, 3, adtl::adouble> &);
 
 template std::array<std::pair<adtl::adouble, Tensor<1, 1, adtl::adouble>>, 1>
-eigenvectors(const SymmetricTensor<2, 1, adtl::adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<1, adtl::adouble>(const SymmetricTensor<2, 1, adtl::adouble> &,
+                               const SymmetricTensorEigenvectorMethod);
 
 template std::array<std::pair<adtl::adouble, Tensor<1, 2, adtl::adouble>>, 2>
-eigenvectors(const SymmetricTensor<2, 2, adtl::adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<2, adtl::adouble>(const SymmetricTensor<2, 2, adtl::adouble> &,
+                               const SymmetricTensorEigenvectorMethod);
 
 template std::array<std::pair<adtl::adouble, Tensor<1, 3, adtl::adouble>>, 3>
-eigenvectors(const SymmetricTensor<2, 3, adtl::adouble> &,
-             const SymmetricTensorEigenvectorMethod);
+eigenvectors<3, adtl::adouble>(const SymmetricTensor<2, 3, adtl::adouble> &,
+                               const SymmetricTensorEigenvectorMethod);
 #endif
 
 // explicit instantiations

--- a/source/base/symmetric_tensor.inst.in
+++ b/source/base/symmetric_tensor.inst.in
@@ -23,9 +23,10 @@ for (deal_II_dimension : DIMENSIONS; number : REAL_SCALARS)
       const SymmetricTensor<2, deal_II_dimension, number> &);
 
     template std::array<std::pair<number, Tensor<1, deal_II_dimension, number>>,
-                        std::integral_constant<int, deal_II_dimension>::value>
-    eigenvectors(const SymmetricTensor<2, deal_II_dimension, number> &,
-                 const SymmetricTensorEigenvectorMethod);
+                        deal_II_dimension>
+    eigenvectors<deal_II_dimension, number>(
+      const SymmetricTensor<2, deal_II_dimension, number> &,
+      const SymmetricTensorEigenvectorMethod);
   }
 
 for (deal_II_dimension : DIMENSIONS; number : DIFFERENTIABLE_ADOLC_REAL_SCALARS)
@@ -46,9 +47,10 @@ for (deal_II_dimension : DIMENSIONS;
       const SymmetricTensor<2, deal_II_dimension, number> &);
 
     template std::array<std::pair<number, Tensor<1, deal_II_dimension, number>>,
-                        std::integral_constant<int, deal_II_dimension>::value>
-    eigenvectors(const SymmetricTensor<2, deal_II_dimension, number> &,
-                 const SymmetricTensorEigenvectorMethod);
+                        deal_II_dimension>
+    eigenvectors<deal_II_dimension, number>(
+      const SymmetricTensor<2, deal_II_dimension, number> &,
+      const SymmetricTensorEigenvectorMethod);
   }
 
 for (deal_II_dimension : DIMENSIONS; number : COMPLEX_SCALARS)

